cms/apps/media/admin.py

```diff
@@ -159,6 +161,88 @@ def get_title(self, obj):
     get_title.admin_order_field = 'title'
     get_title.short_description = 'title'
 
+    def get_related_pages(self, querysets):
+        pages = []
+        for queryset in querysets:
+            for model in queryset:
+                if hasattr(model, 'page'):
+                    pages.append(model)
+                else:
+                    pages.append(model.__class__.objects.get(pk=model.id))
+
+        return pages
+
+    def get_admin_url(self, model):
```

Comment:
`model` is not a good name here - a model in Django speak is a class of objects. Django convention is to use `obj` or `instance`.

This seems too dependent on specific apps being installed. It also seems very long. You don't need to do all this guesswork anyway; the entirety of this function could be replaced with something like this (untested):

```
try:
    return reverse(
        f'admin:{obj._meta.app_label}_{obj._meta.model_name}_change',
        args=[model.pk]
    )
except NoReverseMatch:
    return None
```

Reply:
This is much more concise, flexible and works really well for most cases. However, it doesn't return a valid match for Logos or ContentSections.

For example, trying to get the admin url for a Logo will return 'components/logo/6/change/' when the correct url should actually be 'components/logoset/1/change/‘, because you need to find the logo set that it’s used on.

With these kind of edge cases, I think it's probably acceptable to omit Logo and ContentSection pages unless you have any suggestions on how to handle these kinds of cases.

Reply:
You could try something like this:
```
from cms.apps.pages.admin import page_admin
from django.apps import apps

def get_admin_url(obj, pk):
    try:
        return reverse(
            f'admin:{obj._meta.app_label}_{obj._meta.model_name}_change',
            args=[obj.pk]
        )
    except NoReverseMatch:
        return None

# If it has an associated ModelAdmin we can try and get the admin change page for this object.
if page_admin.site.is_registered(obj):
    return get_admin_url(obj)

# If it does, we try to get the parent model.
try:
    set_obj = apps.get_model(obj._meta.app_label, f'{obj._meta.model_name}Set')
except LookupError:
    set_obj = None

if set_obj:
    # We need to get the relation to the set_obj to find it's pk
    for field in obj._meta.fields:
        # Check if the field is a ForeignKey as most inlines will just be a ForeignKey
        if field.get_internal_type() in ['ForeignKey']:
            # If we find one, check the models it's FK'd to exists and is of the correct type
            related_model = obj._meta.get_field(field).rel.to
            if related_model and isinstance(related_model, set_obj):
                # Naively assume there will oonly be one FK to the parent set model.
                return get_admin_url(related_model)

# Special case OSM section app
try:
    section_base = apps.get_model('sections', 'SectionBase')
except LookupError:
    # If they don't have the sections app installed, bail out here
    return None

# Check if we're a subclass of the SectionBase model
if issubclass(obj, section_base)
    try:
        return get_admin_url(obj.page)
    except AttributeError:
        return None
```
It'll check if a model named xSet exists then find the ForeignKey to it for the original model. If that fails it'll check if the model is a subclass of `SectionBase` and if so return the `page` attribute. I don't want to blindly follow the `page` attribute as it could be used elsewhere. Of course this does assume that the parent for sections will always be called `page` anyway.

Reply:
I am not big on that solution; it's introducing a dependency from the CMS (which should be agnostic) into project-specific models.

I would be extremely surprised if it was impossible to poke around the admin registry (and CMS inlines registered with `register_content_inline`) for `ModelAdmin`-derivatives that have the model registered as an inline.

Reply:
You could iterate over `admin_site._register` to find all the inlines and check if they use the model. Like:
```
for registered_model in admin_site._register:
    if registered_model.inlines:
        for inline in registered_model.inlines:
            if inline.model == obj:
                try:
                    return reverse(
                        f'admin:{registered_model._meta.app_label}_{registered_model._meta.model_name}_change',
                        args=[registered_model.pk]
                    )
                except NoReverseMatch:
                    pass
return None```

Reply:
I've come up with a pretty ugly way of correctly getting the logoset admin urls.

I've tried a bunch of different ways to get the correct URL for ContentSections, but there's not really a way I can think of to get it without doing a check for the 'page' attribute which feels wrong. Anyone have any suggestions?